File: server/src/controller/community.controller.ts
import {
  Controller,
  Post,
  Put,
  Del,
  Get,
  Body,
  Query,
  Inject,
  Param,
} from "@midwayjs/core";
import { Context } from "@midwayjs/koa";
import { ApiTags, ApiOperation } from "@midwayjs/swagger";
import { CommunityService } from "../service/community.service";
import { ResponseUtil } from "../common/types/response";
import { IUserContext } from "../interface";
import { Public } from "../decorator/public.decorator";
import { CreatePostDTO, CreateCommentDTO } from "../dto/community.dto";
import { PAGINATION } from "../common/constants";

@ApiTags("community")
@Controller("/api")
export class CommunityController {
  @Inject()
  communityService!: CommunityService;

  @Public()
  @Get("/posts")
  @ApiOperation({ summary: "游记列表" })
  async getPosts(
    @Query("keyword") keyword?: string,
    @Query("topicId") topicId?: string,
    @Query("userId") userId?: string,
    @Query("page") page?: string,
    @Query("pageSize") pageSize?: string,
  ) {
    const parsedPage = Math.max(1, parseInt(page || "1", 10) || 1);
    const parsedPageSize = Math.min(
      PAGINATION.MAX_PAGE_SIZE,
      Math.max(
        1,
        parseInt(pageSize || String(PAGINATION.DEFAULT_PAGE_SIZE), 10) ||
          PAGINATION.DEFAULT_PAGE_SIZE,
      ),
    );

    const result = await this.communityService.findPosts({
      keyword,
      topicId: topicId ? parseInt(topicId, 10) : undefined,
      userId: userId ? parseInt(userId, 10) : undefined,
      page: parsedPage,
      pageSize: parsedPageSize,
    });

    return ResponseUtil.paginate(
      result.list,
      result.total,
      result.page,
      result.pageSize,
    );
  }

  @Public()
  @Get("/posts/:id")
  @ApiOperation({ summary: "游记详情" })
  async getPostDetail(@Param("id") id: string) {
    const post = await this.communityService.findPostById(parseInt(id, 10));
    if (!post) return ResponseUtil.error("游记不存在", 404);
    return ResponseUtil.success(post);
  }

  @Post("/posts")
  @ApiOperation({ summary: "发布游记" })
  async createPost(ctx: Context, @Body() body: CreatePostDTO) {
    const user = ctx.state.user as IUserContext;
    const post = await this.communityService.createPost({
      userId: user.userId,
      ...body,
    });
    return ResponseUtil.success(post, "发布成功，等待审核");
  }

  @Put("/posts/:id")
  @ApiOperation({ summary: "编辑游记" })
  async updatePost(
    ctx: Context,
    @Param("id") id: string,
    @Body() body: CreatePostDTO,
  ) {
    const user = ctx.state.user as IUserContext;
    const post = await this.communityService.updatePost(
      parseInt(id, 10),
      user.userId,
      body,
    );
    return ResponseUtil.success(post, "更新成功");
  }

  @Del("/posts/:id")
  @ApiOperation({ summary: "删除游记" })
  async deletePost(ctx: Context, @Param("id") id: string) {
    const user = ctx.state.user as IUserContext;
    await this.communityService.deletePost(parseInt(id, 10), user.userId);
    return ResponseUtil.success(null, "已删除");
  }

  @Post("/posts/:id/like")
  @ApiOperation({ summary: "点赞/取消点赞" })
  async togglePostLike(ctx: Context, @Param("id") id: string) {
    const user = ctx.state.user as IUserContext;
    const result = await this.communityService.toggleLike(
      user.userId,
      "POST",
      parseInt(id, 10),
    );
    return ResponseUtil.success(result);
  }

  @Public()
  @Get("/posts/:id/comments")
  @ApiOperation({ summary: "评论列表" })
  async getComments(
    @Param("id") id: string,
    @Query("page") page?: string,
    @Query("pageSize") pageSize?: string,
  ) {
    const parsedPage = Math.max(1, parseInt(page || "1", 10) || 1);
    const parsedPageSize = Math.min(
      PAGINATION.MAX_PAGE_SIZE,
      Math.max(
        1,
        parseInt(pageSize || String(PAGINATION.DEFAULT_PAGE_SIZE), 10) ||
          PAGINATION.DEFAULT_PAGE_SIZE,
      ),
    );

    const result = await this.communityService.findComments(
      parseInt(id, 10),
      parsedPage,
      parsedPageSize,
    );

    return ResponseUtil.paginate(
      result.list,
      result.total,
      result.page,
      result.pageSize,
    );
  }

  @Post("/posts/:id/comments")
  @ApiOperation({ summary: "发表评论" })
  async createComment(
    ctx: Context,
    @Param("id") id: string,
    @Body() body: CreateCommentDTO,
  ) {
    const user = ctx.state.user as IUserContext;
    const comment = await this.communityService.createComment({
      postId: parseInt(id, 10),
      userId: user.userId,
      content: body.content,
      parentId: body.parentId,
      replyToUserId: body.replyToUserId,
    });
    return ResponseUtil.success(comment, "评论成功");
  }

  @Post("/comments/:id/like")
  @ApiOperation({ summary: "评论点赞" })
  async toggleCommentLike(ctx: Context, @Param("id") id: string) {
    const user = ctx.state.user as IUserContext;
    const result = await this.communityService.toggleLike(
      user.userId,
      "COMMENT",
      parseInt(id, 10),
    );
    return ResponseUtil.success(result);
  }

  @Public()
  @Get("/topics")
  @ApiOperation({ summary: "话题列表" })
  async getTopics(
    @Query("page") page?: string,
    @Query("pageSize") pageSize?: string,
  ) {
    const parsedPage = Math.max(1, parseInt(page || "1", 10) || 1);
    const parsedPageSize = Math.min(
      PAGINATION.MAX_PAGE_SIZE,
      Math.max(
        1,
        parseInt(pageSize || String(PAGINATION.DEFAULT_PAGE_SIZE), 10) ||
          PAGINATION.DEFAULT_PAGE_SIZE,
      ),
    );

    const result = await this.communityService.findTopics(
      parsedPage,
      parsedPageSize,
    );
    return ResponseUtil.paginate(
      result.list,
      result.total,
      result.page,
      result.pageSize,
    );
  }

  @Post("/users/:id/follow")
  @ApiOperation({ summary: "关注/取消关注" })
  async toggleFollow(ctx: Context, @Param("id") id: string) {
    const user = ctx.state.user as IUserContext;
    const result = await this.communityService.toggleFollow(
      user.userId,
      parseInt(id, 10),
    );
    return ResponseUtil.success(result);
  }
}

File: server-java/src/main/java/com/wudong/controller/CommunityController.java
package com.wudong.controller;

import com.wudong.common.response.ApiResponse;
import com.wudong.dto.community.CreateCommentRequest;
import com.wudong.dto.community.CreatePostRequest;
import com.wudong.entity.Comment;
import com.wudong.entity.Post;
import com.wudong.entity.Topic;
import com.wudong.security.SecurityUtils;
import com.wudong.service.CommunityService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Tag(name = "社区")
@RestController
@RequiredArgsConstructor
public class CommunityController {

    private final CommunityService communityService;

    @Operation(summary = "获取帖子列表")
    @GetMapping("/api/posts")
    public ApiResponse<Page<Post>> getPosts(
            @RequestParam(required = false) String keyword,
            @RequestParam(required = false) String topicId,
            @RequestParam(required = false) Long userId,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize) {
        return ApiResponse.success(communityService.getPosts(keyword, topicId, userId, PageRequest.of(page - 1, pageSize)));
    }

    @Operation(summary = "获取帖子详情")
    @GetMapping("/api/posts/{id}")
    public ApiResponse<Post> getPost(@PathVariable Long id) {
        return ApiResponse.success(communityService.getPostById(id));
    }

    @Operation(summary = "创建帖子")
    @PostMapping("/api/posts")
    public ApiResponse<Post> createPost(@Valid @RequestBody CreatePostRequest request) {
        Long userId = SecurityUtils.getCurrentUserId();
        Post post = communityService.createPost(userId, request.getTitle(), request.getContent(),
                request.getImages(), request.getVideoUrl(), request.getLocationId(),
                request.getLocationType(), request.getLocationName(), request.getTopicIds());
        return ApiResponse.success(post, "发布成功");
    }

    @Operation(summary = "编辑帖子")
    @PutMapping("/api/posts/{id}")
    public ApiResponse<Post> updatePost(@PathVariable Long id, @Valid @RequestBody CreatePostRequest request) {
        Long userId = SecurityUtils.getCurrentUserId();
        Post post = communityService.updatePost(id, userId, request.getTitle(), request.getContent(),
                request.getImages(), request.getVideoUrl());
        return ApiResponse.success(post, "更新成功");
    }

    @Operation(summary = "删除帖子")
    @DeleteMapping("/api/posts/{id}")
    public ApiResponse<Void> deletePost(@PathVariable Long id) {
        Long userId = SecurityUtils.getCurrentUserId();
        communityService.deletePost(id, userId);
        return ApiResponse.success(null, "删除成功");
    }

    @Operation(summary = "点赞/取消点赞帖子")
    @PostMapping("/api/posts/{id}/like")
    public ApiResponse<Boolean> toggleLikePost(@PathVariable Long id) {
        Long userId = SecurityUtils.getCurrentUserId();
        boolean liked = communityService.toggleLikePost(id, userId);
        return ApiResponse.success(liked);
    }

    @Operation(summary = "获取评论列表")
    @GetMapping("/api/posts/{id}/comments")
    public ApiResponse<Page<Comment>> getComments(
            @PathVariable Long id,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize) {
        return ApiResponse.success(communityService.getComments(id, PageRequest.of(page - 1, pageSize)));
    }

    @Operation(summary = "创建评论")
    @PostMapping("/api/posts/{id}/comments")
    public ApiResponse<Comment> createComment(@PathVariable Long id, @Valid @RequestBody CreateCommentRequest request) {
        Long userId = SecurityUtils.getCurrentUserId();
        Comment comment = communityService.createComment(id, userId, request.getContent(),
                request.getParentId(), request.getReplyToUserId());
        return ApiResponse.success(comment, "评论成功");
    }

    @Operation(summary = "点赞/取消点赞评论")
    @PostMapping("/api/comments/{id}/like")
    public ApiResponse<Boolean> toggleLikeComment(@PathVariable Long id) {
        Long userId = SecurityUtils.getCurrentUserId();
        boolean liked = communityService.toggleLikeComment(id, userId);
        return ApiResponse.success(liked);
    }

    @Operation(summary = "获取话题列表")
    @GetMapping("/api/topics")
    public ApiResponse<Page<Topic>> getTopics(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize) {
        return ApiResponse.success(communityService.getTopics(PageRequest.of(page - 1, pageSize)));
    }

    @Operation(summary = "关注/取消关注用户")
    @PostMapping("/api/users/{id}/follow")
    public ApiResponse<Boolean> toggleFollow(@PathVariable Long id) {
        Long userId = SecurityUtils.getCurrentUserId();
        boolean followed = communityService.toggleFollow(userId, id);
        return ApiResponse.success(followed);
    }
}
